services/horizon/internal/health.go

```diff
@@ -0,0 +1,86 @@
+package horizon
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/stellar/go/protocols/stellarcore"
+	"github.com/stellar/go/support/clock"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/log"
+)
+
+type stellarCoreClient interface {
+	Info(ctx context.Context) (*stellarcore.InfoResponse, error)
+}
+
+type healthCache struct {
+	response   healthResponse
+	lastUpdate time.Time
+	ttl        time.Duration
+	clock      clock.Clock
+	lock       sync.Mutex
+}
+
+func (h *healthCache) get(runCheck func() healthResponse) healthResponse {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if h.clock.Now().Sub(h.lastUpdate) > h.ttl {
+		h.response = runCheck()
+		h.lastUpdate = h.clock.Now()
+	}
+
+	return h.response
+}
+
+func newHealthCache(ttl time.Duration) *healthCache {
+	return &healthCache{ttl: ttl}
+}
+
+type healthCheck struct {
+	session db.SessionInterface
+	ctx     context.Context
+	core    stellarCoreClient
+	cache   *healthCache
+}
+
+type healthResponse struct {
+	DatabaseConnected bool `json:"database_connected"`
+	CoreUp            bool `json:"core_up"`
+	CoreSynced        bool `json:"core_synced"`
+}
+
+func (h healthCheck) runCheck() healthResponse {
+	response := healthResponse{
+		DatabaseConnected: true,
+		CoreUp:            true,
+		CoreSynced:        true,
+	}
+	if err := h.session.Ping(); err != nil {
+		log.Warnf("could ping db: %s", err)
+		response.DatabaseConnected = false
+	}
+	if resp, err := h.core.Info(h.ctx); err != nil {
+		log.Warnf("request to stellar core failed: %s", err)
+		response.CoreUp = false
+		response.CoreSynced = false
+	} else {
+		response.CoreSynced = resp.IsSynced()
+	}
+
+	return response
+}
+
+func (h healthCheck) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	response := h.cache.get(h.runCheck)
+
+	if !response.DatabaseConnected || !response.CoreSynced || !response.CoreUp {
+		w.WriteHeader(http.StatusServiceUnavailable)
+	}
+
+	json.NewEncoder(w).Encode(response)
```

Comment:
We should handle the returned error here just in case.
```suggestion
	err := json.NewEncoder(w).Encode(response)
	if err != nil {
		// ...
	}
```

Reply:
how should we handle it?

Reply:
Just log it and return it `500` HTTP error.

Reply:
in some cases we won't be able to respond with a 500. For example, if we already invoked `w.WriteHeader(http.StatusServiceUnavailable)` we cannot call `w.WriteHeader()` again to change the response code. 

`json.NewEncoder(w).Encode(response)` could also return an error for potentially 2 reasons: 

1) the json serialization failed 
2) writing to the `http.ResponseWriter` failed at some point

If case 2 occurs then it is possible we won't be able to respond with a 500 because the response writer will set the status to 200 before writing anything unless `w.WriteHeader()` has already been called

should we try to account for all these cases to try and respond with a 500 while avoiding multiple response.WriteHeader calls? or is just logging the error sufficient?

Reply:
Good points, I think it's OK if we just log the error. My main concern was ignoring the return error.